Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,628 @@
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request)
+
+        let provider: HTTP1ConnectionProvider = self.connectionProvidersLock.withLock {
+            if let existing = self.connectionProviders[key] {
+                existing.stateLock.withLock {
+                    existing.state.pending += 1
+                }
+                return existing
+            } else {
+                let http1Provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, parentPool: self)
+                self.connectionProviders[key] = http1Provider
+                http1Provider.stateLock.withLock {
+                    http1Provider.state.pending += 1
+                }
+                return http1Provider
+            }
+        }
+
+        return provider.getConnection(preference: preference)
+    }
+
+    func release(_ connection: Connection) {
+        let connectionProvider = self.connectionProvidersLock.withLock {
+            self.connectionProviders[connection.key]
+        }
+        if let connectionProvider = connectionProvider {
+            connectionProvider.release(connection: connection)
+        }
+    }
+
+    func prepareForClose() {
+        let connectionProviders = self.connectionProvidersLock.withLock { self.connectionProviders.values }
+        for connectionProvider in connectionProviders {
+            connectionProvider.prepareForClose()
+        }
+    }
+
+    func syncClose() {
+        let connectionProviders = self.connectionProvidersLock.withLock { self.connectionProviders.values }
+        for connectionProvider in connectionProviders {
+            connectionProvider.syncClose()
+        }
+        self.connectionProvidersLock.withLock {
+            assert(self.connectionProviders.count == 0, "left-overs: \(self.connectionProviders)")
+        }
+    }
+
+    var connectionProviderCount: Int {
+        return self.connectionProvidersLock.withLock {
+            self.connectionProviders.count
+        }
+    }
+
+    /// Used by the `ConnectionPool` to index its `HTTP1ConnectionProvider`s
+    ///
+    /// A key is initialized from a `URL`, it uses the components to derive a hashed value
+    /// used by the `connectionProviders` dictionary to allow retreiving and creating
+    /// connection providers associated to a certain request in constant time.
+    struct Key: Hashable {
+        init(_ request: HTTPClient.Request) {
+            switch request.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            case "unix":
+                self.scheme = .unix
+                self.unixPath = request.url.path
```

Comment:
This doesn't do what you want: to meaningfully construct a UNIX socket URL requires using `URL(string:relativeTo:)`, because it's not possible to have a URL with _two_ path components in it. Observe:

```
  3> let socketURL = URL(string: "unix:///var/run/docker.sock")!
socketURL: URL = {}
  4> let finalURL = URL(string: "/users/list", relativeTo: socketURL)!
finalURL: URL = {}
  5> print(finalURL)
/users/list -- unix:///var/run/docker.sock
  6> print(finalURL.host)
nil
  7> print(finalURL.path)
/users/list
  8> print(finalURL.absoluteURL)
unix:///users/list
```

I think the rule here is to do the same as the higher level code did: `request.url.baseURL.path`. That should give you the path to the unix socket.

Reply:
Fixed thanks to https://github.com/swift-server/async-http-client/pull/105/commits/c2355f67d4540a1234c9b041cd788cb9f5809ec9